tests/model_connectors/test_github_conn.py

```diff
@@ -0,0 +1,143 @@
+import base64
+import pytest
+
+from garden_ai.model_connectors import GitHubConnector
+from garden_ai.model_connectors.exceptions import (
+    ConnectorLFSError,
+    ConnectorInvalidRepoIdError,
+)
+
+
+def test_GHconnector_stage_is_idempotent(
+    mocker,
+    patch_has_lfs,
+    patch_fetch_readme,
+):
+    # Mock os.path.exists and os.mkdir
+    mocker.patch("os.path.exists", return_value=True)
+    mocker.patch("os.mkdir")
+
+    # Mock Repo to simulate both clone and pull scenarios
+    mock_repo_class = mocker.patch("garden_ai.model_connectors.github_conn.Repo")
+    mock_repo_base = mocker.patch("garden_ai.model_connectors.model_connector.Repo")
+    mock_repo_instance = mocker.MagicMock()
+    mock_repo_class.return_value = mock_repo_instance
+    mock_repo_base.return_value = mock_repo_instance
+    mock_repo_instance.remotes.origin.pull = mocker.MagicMock()
+    # Set up the mock to return a URL that matches the connector's repo_url
+    mock_repo_instance.remotes.origin.url = "https://github.com/fake/repo.git"
+
+    # Mock Repo.clone_from method to track calls without actually cloning
+    mock_clone_from = mocker.patch(
+        "garden_ai.model_connectors.github_conn.Repo.clone_from"
+    )
+
+    # Mock is_git_dir to control the flow in the stage method
+    mocker.patch(
+        "garden_ai.model_connectors.model_connector.is_git_dir",
+        side_effect=[
+            False,
+            True,
+            True,
+        ],  # First call: not a git dir, then it is a git dir
+    )
+
+    fake_commit_hash = "a" * 40
+
+    # enable_imports=False just bc mocking sys.path.append was hard
+    connector = GitHubConnector(
+        repo_url="https://github.com/fake/repo",
+        enable_imports=False,
+        revision=fake_commit_hash,
+    )
+
+    # First call should trigger clone since it's not a git dir yet
+    connector.stage()
+    mock_clone_from.assert_called_once_with(
+        "https://github.com/fake/repo.git", "models/repo", branch="main"
+    )
+
+    # Reset mock to test idempotency on subsequent calls
+    mock_clone_from.reset_mock()
+
+    # Subsequent calls should not trigger clone_from again, but should pull
+    connector.stage()
+    connector.stage()
+
+    # Assert that Repo.clone_from was not called again after the first time
+    mock_clone_from.assert_not_called()
+    # Assert that pull was called on subsequent invocations
+    assert (
+        mock_repo_instance.remotes.origin.pull.call_count == 2
+    ), "Pull should be called on subsequent calls"
+
+
+def test_GitHubConnector_pins_commit_hash_correctly(
+    mocker,
+    patch_has_lfs,
+    patch_fetch_readme,
+):
+    # Mock the response from the GitHub API
+    fake_commit_hash = "a" * 40
```

Comment:
every time I see this I read it in my head like the first bit of immigrant song lol